core-util/SharedPointer.h

```diff
@@ -0,0 +1,275 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __SHAREDPOINTER_H__
+#define __SHAREDPOINTER_H__
+
+#include "core-util/core-util.h"
+#include "mbed-alloc/ualloc.h"
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifndef NDEBUG
+#include <stdio.h>
+#define SHAREDPOINTER_DEBUG(...) { printf(__VA_ARGS__); }
```

Comment:
In the future, we might want to add a `CORE_UTIL_DEBUG` macro in `core-util.h`. For now, this is fine.